airbyte-workers/src/main/java/io/airbyte/workers/RecordSchemaValidator.java

```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright (c) 2021 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.workers;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import io.airbyte.config.StandardSyncInput;
+import io.airbyte.protocol.models.AirbyteRecordMessage;
+import io.airbyte.validation.json.JsonSchemaValidator;
+import io.airbyte.validation.json.JsonValidationException;
+import java.util.HashMap;
+
+public class RecordSchemaValidator {
+
+  private final HashMap<String, JsonNode> streams;
+
+  public RecordSchemaValidator(final StandardSyncInput syncInput) {
+    final String streamPrefix = syncInput.getPrefix();
+
+    // streams is populated with a stream name (including prefix) and stream schema
+    // for easy access when we check each record's schema
+    this.streams = new HashMap<String, JsonNode>();
+    syncInput.getCatalog().getStreams().forEach(s -> {
+      streams.put(String.format(streamPrefix + s.getStream().getName().trim()), s.getStream().getJsonSchema());
+    });
```

Comment:
I had suggested adding this logic of creating a map of stream name to stream schema in the constructor so that we didn't have to filter the list of streams on every `validateSchema` call, which you did correctly here.

However, looking back on this I realize that we probably should not have somewhat complex logic and the creation of objects like this inside of constructors, since I believe that is sort of an antipattern in Java.

@jdpgrailsdev do you have any thoughts here on how we should approach this? I think there is still value in creating this map of stream name -> stream record schema up front so that the validation calls themselves are very quick, but perhaps this should be done in a factory class/method instead? What is the best approach with the future DI work in mind?

Reply:
@lmossman I am missing some context here about the `StandardSyncInput`, catalogs and streams (how/where they are created, etc).  Generally speaking, I would agree that in the DI world, this map would be something that we would want to create once and inject as a bean.  Of course, that only works if the map can be created separate from this code/can be constructed at startup.  In the meantime, any change that we could make that moves to a "create once" approach would be best and would make the transition to DI easier.  Maybe this code could be encapsulated in the `StandardSyncInput` so that it is built when that is created and saved as a member and exposed via method.

Sort of related, that lambda could be re-written to use `.collect(Collectors.toMap(k -> String.format(streamPrefix + k.getStream().getName().trim()), v -> v.getStream().getJsonSchema())`

Reply:
Following this all the way up the chain, the `StandardSyncInput` is first constructed [here](https://github.com/airbytehq/airbyte/blob/fd7f21e07170fbc9cb6a22d03b642486d0da0b52/airbyte-workers/src/main/java/io/airbyte/workers/temporal/scheduling/activities/GenerateInputActivityImpl.java#L59-L70), and then is passed down through a bunch of classes until it is eventually serialized and written to a file on the orchestrator pod. This pod then deserializes that file back into an object [here](https://github.com/airbytehq/airbyte/blob/fd7f21e07170fbc9cb6a22d03b642486d0da0b52/airbyte-container-orchestrator/src/main/java/io/airbyte/container_orchestrator/ReplicationJobOrchestrator.java#L59), where it is then passed into the DefaultReplicationWorker [here](https://github.com/airbytehq/airbyte/blob/fd7f21e07170fbc9cb6a22d03b642486d0da0b52/airbyte-container-orchestrator/src/main/java/io/airbyte/container_orchestrator/ReplicationJobOrchestrator.java#L102). 

Would a good intermediate change be to add this map creation logic to the ReplicationJobOrchestrator after the object is deserialized, and then pass that map into this RecordSchemaValidationException class when it is constructed in that method?

Reply:
I think it will be hard to encapsulate this code in `StandardSyncInput`, because that is just a POJO that gets generated from this json schema: https://github.com/airbytehq/airbyte/blob/master/airbyte-config/models/src/main/resources/types/StandardSyncInput.yaml

I think it will be a pretty ugly change to try to add this map all the way up the call chain to the same place that the StandardSyncInput is first constructed, where I linked above. I don't want to block this PR much longer on this decision, so maybe it makes sense to just leave this as-is for now. Though @alovew I think we should rewrite lambda function as Jonathan proposed above, which will allow the `this.streams` initalization to all happen on a single line

Reply:
@lmossman The general dependency injection guidance would be that objects should not create object internally.  For instance, if class A currently instantiates an instance of class B in class A's constructor, we would want to change this so that class B is created by whatever common code also instantiates class A and passes the B instance into A (in the future, this would be converted to use auto-wiring).  So, in this case, if there is a higher class that creates the `RecordSchemaValidator` could also create this map and pass it to the `RecordSchemaValidator` via its constructor.  This would allow us to eventually move the creation of the map to the dependency injection configuration, if possible.

Reply:
Got it, then sounds like we should probably go with what I suggested in my first reply to you above and create the map outside of the RecordSchemaValidator, then pass it into `RecordSchemaValidator`. There are currently two different places where the RecordSchemaValidator is constructed (ReplicationJobOrchestrator and ReplicationActivityImpl) so we will probably want to have some sort of utility/library function to create this map from the standard sync input, or a factory class for the ReplicationActivityImpl that does this. I'll leave that decision up to @alovew as the implementer of these chanes